baml_language/crates/baml_thir/src/lib.rs

```diff
@@ -751,13 +751,387 @@ fn infer_expr<'db>(ctx: &mut TypeContext<'db>, expr_id: ExprId, body: &ExprBody)
             }
         }
 
+        Expr::Match { scrutinee, arms } => {
```

Comment:
Worth putting a comment at the top of this branch: `// Match expressions synthesize a type`.

At some point, we should also add a function that checks matches, because bidirectional typing switches from synthesis to checking in certain circumstances:

```
/// This should _check_ the match branches, rather than synthesizing.
let x: int = match (true) {
  true => 0,
  false => 1,
}